### ISSUE TITLE
False positive in MemorySanitizer

### ISSUE THREAD

**Issue by shao-hua-li** [llvm/llvm-project#61946]
For the following code, `clang -O1 -fsanitize=memory` reports a use-of-uninitialized memory, but there is no in fact. `clang -O0 -fsanitize=memory` emits nothing.

Compiler explorer: https://godbolt.org/z/qaTdEb1rb

```console
% cat a.c
static int b = -2;
int c;
int d ;
short e;
void main() {
    if ((d |= e) << b)
        c = 5;
    int *f[5] = {&b};
}
%
% clang -O0 -fsanitize=memory a.c && ./a.out
%
% clang -O1 -fsanitize=memory a.c && ./a.out
==398997==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x55f596212423 in main /a.c:6:9
...
%
```

**Comment by xgupta** [comment]
FYI, instCombinePass removing the alloc instruction, and GlobalOptPass removing the load instruction before MemotySanitizerPass which might be removing a code path that initializes the memory. See Opt pipeline for -O1 - https://godbolt.org/z/qsETbrT4c.

**Comment by nikic** [comment]
The code contains undefined behavior (shift by negative amount). It is correctly detected by `-fsanitize=undefined`.

**Comment by xgupta** [comment]
Okay, but there is a difference b/w -O0 and -O1 behavior.
